python-package/xgboost/core.py

```diff
@@ -138,27 +138,42 @@ def c_array(ctype, values):
     return (ctype * len(values))(*values)
 
 
-def _maybe_from_pandas(data, feature_names, feature_types):
+def _maybe_from_pandas(data, label, feature_names, feature_types):
     """ Extract internal data from pd.DataFrame """
     try:
         import pandas as pd
     except ImportError:
-        return data, feature_names, feature_types
+        return data, label, feature_names, feature_types
 
     if not isinstance(data, pd.DataFrame):
-        return data, feature_names, feature_types
-
-    dtypes = data.dtypes
-    if not all(dtype.name in ('int64', 'float64', 'bool') for dtype in dtypes):
-        raise ValueError('DataFrame.dtypes must be int, float or bool')
+        return data, label, feature_names, feature_types
+
+    mapper = {'int8': 'int', 'int16': 'int', 'int32': 'int', 'int64': 'int',
+              'uint8': 'int', 'uint16': 'int', 'uint32': 'int', 'uint64': 'int',
+              'float16': 'float', 'float32': 'float', 'float64': 'float',
+              'bool': 'i'}
+
+    data_dtypes = data.dtypes
+    if not all(dtype.name in (mapper.keys()) for dtype in data_dtypes):
+        raise ValueError('DataFrame.dtypes for data must be int, float or bool')
+
+    if label is not None:
+        if isinstance(label, pd.DataFrame):
```

Comment:
I understand label must be 1-d array. Thus:

- Should check ``label`` is either ``Series`` or single-column ``DataFrame``, then check ``dtype``. 
- If label Is 1-d ``DataFrame`` values must be flatten (it's nice to have a test).